### PR TITLE
List patterns: Slice value is assumed to be never null

### PR DESCRIPTION
**Pull request by alrz**
Spec is updated as follows:

>The member that makes the type sliceable is assumed to be well-behaved, that is, the return value is never null and that it is a proper subslice of the containing list.

~~In a nullable-enabled context, the annotation on the output type is ignored to avoid producing nonsensical warnings.~~
~~(no longer needed to ignore, we're emitting the test but assuming the false branch is unreachable)~~ not an option.

Updated LDM decision (ended up agreeing with this approach): https://github.com/dotnet/csharplang/issues/5599

Test plan: https://github.com/dotnet/roslyn/issues/51289

### CODE REVIEW

**Comment by jcouv** [comment]
I'm not following. The LDM decision was to honor the annotation on Slice method. The current behavior is correct, no?

**Comment by alrz** [comment]
@jcouv

Note that this is an unconfirmed change.

The current behavior is correct but it doesn't work with subsumption checking with nested slice (see Slice_00 with arrays)

I'm proposing to drop the null check on the slice value:

- That would only affect the codegen and nullability exhaustiveness (won't give "impossible match" for `[..null]`)
- If the type is annotated (which is not what we'd expect as "well-behaved"), a slice variable remains annotated.

The second point is what's been discussed and this PR doesn't change that.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (alrz) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:2719]: Both warn about dereferenced of a possible null because we never check the value for null. 
This type is malformed under our well-behaved assumption, so the output is nonsensical.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTestBase.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTestBase.cs:444]: What's the purpose of making this method conditional?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:2624]:
```suggestion
            rest2.Value.ToString(); // (no warning)
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (alrz) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:2679]: probably should check the annotation instead of state but the original warning still can't be avoided e.g. with `{} rest`.
what is the expected behavior here?

Comment (alrz) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:2679]: (`T` is oblivious, I think `where T : notnull` should make it go away.)

File: src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs:426]: Here's the current behavior: it is safe (no null-ref exception thrown because we do emit a null check) but annoying (an arguably incorrect subsumption error). I think that's an acceptable behavior.

```
#nullable enable

if (new C() is [_, ..[0, _]]) // doesn't crash (IL is correct)
{
    
}

_ = new C() switch
{
   [_, 0, _] => 0,
   [_, ..[0, _]] => 0, // error CS8510: The pattern is unreachable.
   _ => 0,
};

class C
{
    public int Length => 3;
    public int this[int i] => 0;
    public C? Slice(int i, int j) => null;
}
```

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs:426]: Let's add a test for the scenario you pointed out offline (difference between `Span<int>` and `int[]`) [sharplab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQBsCWyaoA6wwQIB2yAPgAIBMAjALABQAygSQDyYkwB8ABMj4BePgBMIYKHHQwA3E2QB3TDADGACz4AKZAEo+AbyZ8TfVVGQQ+AbQB0t67QC6AGmouAzE5DHT5yza0bp7efAD0YXzECAD2CL6mpsBIUADW8owAvkxMXDDWTnxQwmISUjIZSioa2lD6RoyJ/lZ2Ds7BXj6NfhYtQe6d4ZEkMVEIsfHdiSbJ0OlM2YxAA==)

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (alrz) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:5630]: ❓ Should `e is [.. null]` flagged as an impossible match as well?

**Comment by jcouv** [comment]
From discussion with @333fred we're still leaning towards keeping the current behavior.
Yes, we have an inconsistent diagnostic between `Span<int>` and `int[]` in this [example](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQBsCWyaoA6wwQIB2yAPgAIBMAjALABQAygSQDyYkwB8ABMj4BePgBMIYKHHQwA3E2QB3TDADGACz4AKZAEo+AbyZ8TfVVGQQ+AbQB0t67QC6AGmouAzE5DHT5yza0bp7efAD0YXzECAD2CL6mpsBIUADW8owAvkxMXDDWTnxQwmISUjIZSioa2lD6RoyJ/lZ2Ds7BXj6NfhYtQe6d4ZEkMVEIsfHdiSbJ0OlMmUA).
Yes, users can already get in trouble with mis-behaved types (such as one that returns Length=-1) as in this [example](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQBsCWyaoA6wwQIB2yANDCAnCQD4ACATAIwCwAUMwAQDCnAb07cR3PAkwA3WBG6YSMbuggkA5jAAWAbmGiGAZj4AKeYuVrNASm4BeAHzdN2AHTn1G20pXudHUdwM5BUcNbABtUzkAXVsHFl9/QMiAZSwAYwgTYMxyIMUAK2t7EIQAewB3bhIIcqNLBL1DSIAZb01uAW5VCEVOhhYATiMSKABbCFKwI1aLDUt6gIB2L1mtbgBfDc51zh4ADUFdEUD+gDZuYFLS9G4AFRQYFiNebjSihzS5ZG4w52cohrHQxnC5XG73HBMZ6vd6vL4/P4AfQBRwCTCYqKEfn8ekGRghMCh1VqqBY83qqM22yAA=).
But removing the null check from the DAG leads to a worse kind of problem for users with mis-behaving types.

I think these were the two pivotal examples as part of our discussion, but let me know if I didn't summarize the case you're presenting adequately.

**Comment by alrz** [comment]
> Yes, we have an inconsistent diagnostic between `Span<int>` and `int[]` in this example.

If this is ok (implying we still want to give errors where we already do), then I think no further changes is required here. thanks.
(removal of IDS_FeatureLengthPattern resource string can be done separately)

**Comment by jcouv** [comment]
Following discussion with Mads (who leans towards removing the null check for Slice), I'm going to bring this to LDM again by email. This won't block merging the feature to `main` branch.

**Comment by alrz** [comment]
To be clear, that's as opposed to removing the consistency assumption (that would simply things a lot but I'm fine with both)

Marked as unresolved: https://github.com/dotnet/roslyn/pull/57457#discussion_r746948358 (In case we go with dropping the null test)
Perhaps we could use a different message there.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (alrz) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:3945]: @jcouv thoughts on what diag message would make sense here?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs
Comment (alrz) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:679]: I'm not entirely sure about this check. @RikkiGibson is this the right way do this?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:679]: This check can be removed after all (not necessary). Thanks

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 9)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs
Comment (alrz) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:679]: Will do. I didn't understand what's the rationale though. Some of warnings may end up being nonsensical. Is that ok?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 17)

**Comment by jcouv** [comment]
@333fred @AlekseyTs @dotnet/roslyn-compiler for second review. Thanks

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:3988]: I'm not sure I understand this one. Shouldn't the `[]` pattern say that it's a non-null slice, since it has to dereference the slice to determine if the slice is empty?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:4768]: Original lines and numbers should be updated.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:6765]: I don't understand why were using Random here. Why are we not just testing everything?

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 17)

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (alrz) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:6765]: The original test takes too long even with a pairwise setting. This takes 1-2 sec on my machine. I think that's better than skipping it entirely.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (alrz) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:3988]: Makes sense. I'd defer to @jcouv for that. My solution was to ignore and report the annotation on the return type.

If we consider slice.Length evaluation (or any other) as a dereference with no warning on forthcoming derefs, shouldn't we inform the user about it?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:6765]: I see. Consider documenting it as such.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (alrz) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:3988]: I think we can do something in between: ignore and report only if there's a dereference i.e. when the result is being used - that would exclude var patterns.

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:3988]: Thanks for spotting this. There's indeed a good question here.
From the LDM notes: "We will adjust our expectations around slice patterns to assume that Slice will never return a null value, and we will not insert a null test into the reachability or codegen DAGs."

I think it should follow that both the `.. var slice` and `.. [] slice2` cases should yield not-null state for those variables.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs
Comment (alrz) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:517]: ❓ Should we also strip the annotation on the return type for semantic model?

Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:517]: The `var` will definitely have a `?`, because it always does. Theoretically, though, it should say that `variable` is considered not null here.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:3883]: This isn't quite what I meant: the numbers are wrong, because warning 2 is no longer there.

Comment (alrz) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:3883]: I renumbered the one you commented on. Will update the rest in the next commit.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (alrz) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs]: @jcouv Can you please take another look? thanks.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 18). 
Nit: some line-number comments in tests could be adjusted, if no longer producing a diagnostic

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (alrz) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs]: Done.

**Review by 333fred — COMMENTED** [review]
LGTM (commit 19)

**Review by 333fred — APPROVED** [review]

**Comment by jcouv** [comment]
Thanks @alrz! :-)

**Pull request merged**
